lib/galaxy/jobs/runners/local.py

```diff
@@ -213,12 +225,70 @@ def _check_pid(self, pid):
             return False
 
     def _terminate(self, proc):
+        if proc.container_commands:
+            self._run_command(proc.container_commands['kill'])
         os.killpg(proc.pid, 15)
         sleep(1)
         if proc.poll() is None:
             os.killpg(proc.pid, 9)
         return proc.wait()  # reap
 
+    def _stop_container(self, job_wrapper):
+        return self._run_container_command(job_wrapper, 'stop')
+
+    def _kill_container(self, job_wrapper):
+        return self._run_container_command(job_wrapper, 'kill')
+
+    def _run_container_command(self, job_wrapper, command):
+        job = job_wrapper.get_job()
+        if job:
+            cont = job.container
+            if cont:
+                if cont.container_type == 'docker':
+                    return self._run_command(cont.container_info['commands'][command])
+
+    def _run_command(self, command):
+        exit_code = subprocess.call(command,
+                                    shell=True,
+                                    env=self._environ,
+                                    preexec_fn=os.setpgrp)
+        log.debug('_run_command(%s) exit code %s', command, exit_code)
+        return exit_code
+
+    def __handle_container(self, job_wrapper, proc):
+        proc.container_commands = None
+        job = job_wrapper.get_job()
+        if job:
+            cont = job.container
+            if cont:
+                if cont.container_type == 'docker':
+                    # Handle getting RealTimeTool ports, if applicable
+                    eps = job.realtimetool_entry_points
+                    if eps:
+                        max_command_attempts = 10 + 1
+                        ports_raw = None
+                        for i in range(1, max_command_attempts):
+                            with tempfile.TemporaryFile() as stdout_file:
+                                exit_code = subprocess.call(cont.container_info['commands']['port'],
```

Comment:
We need to capture and parse `stdout` to get the port info here, whereas in the other cases, we just let the stream pass through. Don't want the overhead of making a temp file when we don't need it. Could make it flag-able in the method, but it might be messier due to the with.